src/boot/measure.c

```diff
@@ -464,6 +525,235 @@ static int verb_calculate(int argc, char *argv[], void *userdata) {
+
+                rc = sym_Esys_PolicyPCR(
+                                c.esys_context,
+                                session_handle,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                &intermediate_digest,
+                                &pcr_selection);
+                if (rc != TSS2_RC_SUCCESS) {
+                        r = log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                            "Failed to push PCR policy into TPM: %s", sym_Tss2_RC_Decode(rc));
+                        goto finish;
+                }
+
+                _cleanup_(Esys_Freep) TPM2B_DIGEST *pcr_policy_digest = NULL;
+                rc = sym_Esys_PolicyGetDigest(
+                                c.esys_context,
+                                session_handle,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                &pcr_policy_digest);
+                if (rc != TSS2_RC_SUCCESS) {
+                        r = log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                            "Failed to get policy digest from TPM: %s", sym_Tss2_RC_Decode(rc));
+                        goto finish;
+                }
+
+                session_handle = tpm2_flush_context_verbose(c.esys_context, session_handle);
+
+                _cleanup_(EVP_MD_CTX_freep) EVP_MD_CTX* mdctx = NULL;
+                mdctx = EVP_MD_CTX_new();
+                if (!mdctx) {
+                        r = log_oom();
+                        goto finish;
+                }
+
+                if (EVP_DigestSignInit(mdctx, NULL, p->md, NULL, privkey) != 1) {
+                        r = log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                            "Failed to initialize signature context.");
+                        goto finish;
+                }
+
+                if (EVP_DigestSignUpdate(mdctx, pcr_policy_digest->buffer, pcr_policy_digest->size) != 1) {
+                        r = log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                            "Failed to sign data.");
+                        goto finish;
+                }
+
+                size_t ss;
+                if (EVP_DigestSignFinal(mdctx, NULL, &ss) != 1) {
+                        r = log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                            "Failed to finalize signature");
+                        goto finish;
+                }
+
+                _cleanup_free_ void *sig = malloc(ss);
+                if (!ss) {
+                        r = log_oom();
+                        goto finish;
+                }
+
+                if (EVP_DigestSignFinal(mdctx, sig, &ss) != 1) {
```

Comment:
I'm confused Up here https://github.com/systemd/systemd/pull/24351/files#diff-087f16a06bf226a0017c61f4cdb346ae9a6895d4a18dc5587d30675d7686fafbR88 it says that TPM is needed for signing, but it looks like you sign the digest with OpenSSL.

Reply:
Well, we generate the policy hash via TPM, and then sign the resulting hash with openssl.

If this code is changed to implement the hashing of the policy client side without talking to a TPM we won't need a TPM anymore at all.